File: backend/routers/invoices.py
from fastapi import APIRouter, HTTPException, Query
from typing import List, Optional
from database import get_db_connection
from models import InvoiceCreate, InvoicePaymentUpdate, InvoiceResponse
from datetime import datetime, date

router = APIRouter(prefix="/api/invoices", tags=["Invoices"])

@router.get("/suggest/{room_id}")
def suggest_invoice_data(room_id: int):
    conn = get_db_connection()
    cursor = conn.cursor()
    
    # 1. Tìm hợp đồng đang active
    cursor.execute("""
    SELECT c.*, r.room_code, r.floor, r.position
    FROM contracts c
    JOIN rooms r ON c.room_id = r.id
    WHERE c.room_id = ? AND c.status = 'active'
    ORDER BY c.id DESC LIMIT 1
    """, (room_id,))
    contract = cursor.fetchone()
    if not contract:
        conn.close()
        raise HTTPException(status_code=400, detail="Phòng này hiện chưa có hợp đồng thuê hoạt động!")

    # 2. Tìm chỉ số điện nước hóa đơn gần nhất
    cursor.execute("""
    SELECT curr_electricity, curr_water, billing_month
    FROM invoices
    WHERE room_id = ?
    ORDER BY billing_month DESC, id DESC LIMIT 1
    """, (room_id,))
    last_inv = cursor.fetchone()

    prev_elec = last_inv["curr_electricity"] if last_inv else 0.0
    prev_water = last_inv["curr_water"] if last_inv else 0.0

    today = date.today()
    current_month = today.strftime("%Y-%m")

    result = {
        "room_id": room_id,
        "room_code": contract["room_code"],
        "floor": contract["floor"],
        "position": contract["position"],
        "contract_id": contract["id"],
        "tenant_name": contract["tenant_name"],
        "tenant_phone": contract["tenant_phone"],
        "billing_month": current_month,
        "billing_date": today.strftime("%Y-%m-%d"),
        "prev_electricity": prev_elec,
        "curr_electricity": prev_elec, # Mặc định để người dùng nhập tiếp
        "electricity_rate": contract["electricity_rate"],
        "water_billing_type": contract["water_billing_type"],
        "prev_water": prev_water,
        "curr_water": prev_water,
        "water_rate": contract["water_rate"],
        "water_person_count": contract["water_person_count"],
        "room_amount": contract["rent_price"],
        "internet_amount": contract["internet_fee"],
        "garbage_amount": contract["garbage_fee"],
        "parking_amount": contract["parking_fee"],
        "other_amount": contract["other_fee"],
        "other_note": contract["other_fee_note"] or ""
    }
    conn.close()
    return result

@router.get("", response_model=List[InvoiceResponse])
def get_invoices(
    month: Optional[str] = None,
    room_id: Optional[int] = None,
    payment_status: Optional[str] = None
):
    conn = get_db_connection()
    cursor = conn.cursor()
    query = """
    SELECT 
        inv.*,
        r.room_code,
        r.floor,
        c.tenant_name,
        c.tenant_phone,
        w.name as wallet_name
    FROM invoices inv
    JOIN rooms r ON inv.room_id = r.id
    JOIN contracts c ON inv.contract_id = c.id
    LEFT JOIN wallets w ON inv.wallet_id = w.id
    WHERE 1=1
    """
    params = []
    if month:
        query += " AND inv.billing_month = ?"
        params.append(month)
    if room_id:
        query += " AND inv.room_id = ?"
        params.append(room_id)
    if payment_status:
        query += " AND inv.payment_status = ?"
        params.append(payment_status)

    query += " ORDER BY inv.billing_month DESC, r.floor ASC, r.room_code ASC"
    cursor.execute(query, params)
    rows = cursor.fetchall()
    conn.close()
    return [dict(row) for row in rows]

@router.get("/{invoice_id}", response_model=InvoiceResponse)
def get_invoice(invoice_id: int):
    conn = get_db_connection()
    cursor = conn.cursor()
    cursor.execute("""
    SELECT 
        inv.*,
        r.room_code,
        r.floor,
        c.tenant_name,
        c.tenant_phone,
        w.name as wallet_name
    FROM invoices inv
    JOIN rooms r ON inv.room_id = r.id
    JOIN contracts c ON inv.contract_id = c.id
    LEFT JOIN wallets w ON inv.wallet_id = w.id
    WHERE inv.id = ?
    """, (invoice_id,))
    row = cursor.fetchone()
    conn.close()
    if not row:
        raise HTTPException(status_code=404, detail="Không tìm thấy hóa đơn")
    return dict(row)

@router.post("", response_model=InvoiceResponse)
def create_invoice(inv: InvoiceCreate):
    conn = get_db_connection()
    cursor = conn.cursor()

    try:
        # Tính toán điện
        elec_usage = max(0.0, inv.curr_electricity - inv.prev_electricity)
        elec_amount = elec_usage * inv.electricity_rate

        # Tính toán nước
        if inv.water_billing_type == "meter":
            water_usage = max(0.0, inv.curr_water - inv.prev_water)
            water_amount = water_usage * inv.water_rate
        elif inv.water_billing_type == "per_person":
            water_usage = inv.curr_water # có thể lưu số người
            water_amount = inv.water_amount if inv.water_amount > 0 else inv.water_rate
        else:
            water_usage = 0.0
            water_amount = inv.water_amount

        total = (
            inv.room_amount +
            elec_amount +
            water_amount +
            inv.internet_amount +
            inv.garbage_amount +
            inv.parking_amount +
            inv.other_amount -
            inv.discount_amount
        )

        cursor.execute("""
        INSERT INTO invoices (
            contract_id, room_id, billing_month, billing_date,
            prev_electricity, curr_electricity, electricity_usage, electricity_rate, electricity_amount,
            water_billing_type, prev_water, curr_water, water_usage, water_rate, water_amount,
            room_amount, internet_amount, garbage_amount, parking_amount, other_amount, other_note,
            discount_amount, total_amount, payment_status, paid_amount, paid_date, wallet_id, notes
        ) VALUES (
            ?, ?, ?, ?,
            ?, ?, ?, ?, ?,
            ?, ?, ?, ?, ?, ?,
            ?, ?, ?, ?, ?, ?,
            ?, ?, ?, ?, ?, ?, ?
        )
        """, (
            inv.contract_id, inv.room_id, inv.billing_month, inv.billing_date,
            inv.prev_electricity, inv.curr_electricity, elec_usage, inv.electricity_rate, elec_amount,
            inv.water_billing_type, inv.prev_water, inv.curr_water, water_usage, inv.water_rate, water_amount,
            inv.room_amount, inv.internet_amount, inv.garbage_amount, inv.parking_amount, inv.other_amount, inv.other_note or "",
            inv.discount_amount, total, inv.payment_status, inv.paid_amount, inv.paid_date, inv.wallet_id, inv.notes or ""
        ))
        inv_id = cursor.lastrowid
        conn.commit()

        cursor.execute("""
        SELECT 
            inv.*,
            r.room_code,
            r.floor,
            c.tenant_name,
            c.tenant_phone,
            w.name as wallet_name
        FROM invoices inv
        JOIN rooms r ON inv.room_id = r.id
        JOIN contracts c ON inv.contract_id = c.id
        LEFT JOIN wallets w ON inv.wallet_id = w.id
        WHERE inv.id = ?
        """, (inv_id,))
        row = cursor.fetchone()
        return dict(row)
    except Exception as e:
        conn.rollback()
        raise e
    finally:
        conn.close()

@router.post("/{invoice_id}/pay")
def pay_invoice(invoice_id: int, pay_data: InvoicePaymentUpdate):
    conn = get_db_connection()
    cursor = conn.cursor()

    try:
        cursor.execute("""
        SELECT inv.*, r.room_code, c.tenant_name
        FROM invoices inv
        JOIN rooms r ON inv.room_id = r.id
        JOIN contracts c ON inv.contract_id = c.id
        WHERE inv.id = ?
        """, (invoice_id,))
        inv = cursor.fetchone()
        if not inv:
            raise HTTPException(status_code=404, detail="Không tìm thấy hóa đơn")

        # Cập nhật trạng thái thanh toán hóa đơn
        cursor.execute("""
        UPDATE invoices
        SET payment_status = ?, paid_amount = ?, paid_date = ?, wallet_id = ?
        WHERE id = ?
        """, (pay_data.payment_status, pay_data.paid_amount, pay_data.paid_date, pay_data.wallet_id, invoice_id))

        if pay_data.auto_create_income and pay_data.payment_status == "paid":
            # Tìm category "Thu tiền Nhà trọ (Phòng thuê & Dịch vụ)"
            cursor.execute("SELECT id FROM categories WHERE name LIKE '%Thu tiền Nhà trọ%' LIMIT 1")
            cat_row = cursor.fetchone()
            cat_id = cat_row["id"] if cat_row else None

            # Tạo giao dịch Thu
            title = f"Thu tiền trọ {inv['room_code']} tháng {inv['billing_month']} ({inv['tenant_name']})"
            cursor.execute("""
            INSERT INTO transactions (title, amount, type, category_id, wallet_id, transaction_date, notes, ref_invoice_id)
            VALUES (?, ?, 'income', ?, ?, ?, ?, ?)
            """, (title, pay_data.paid_amount, cat_id, pay_data.wallet_id, f"{pay_data.paid_date} 12:00", f"Tự động tạo từ hóa đơn #{invoice_id}", invoice_id))

            # Cộng tiền vào ví
            cursor.execute("UPDATE wallets SET balance = balance + ? WHERE id = ?", (pay_data.paid_amount, pay_data.wallet_id))

        conn.commit()
        return {"success": True, "message": "Đã thanh toán hóa đơn và đồng bộ vào Quỹ thu chi thành công"}
    except Exception as e:
        conn.rollback()
        raise e
    finally:
        conn.close()

@router.delete("/{invoice_id}")
def delete_invoice(invoice_id: int):
    conn = get_db_connection()
    cursor = conn.cursor()
    # Kiểm tra xem có giao dịch thu nhập nào liên kết không
    cursor.execute("SELECT id, amount, wallet_id FROM transactions WHERE ref_invoice_id = ?", (invoice_id,))
    tx_rows = cursor.fetchall()
    for tx in tx_rows:
        cursor.execute("UPDATE wallets SET balance = balance - ? WHERE id = ?", (tx["amount"], tx["wallet_id"]))
        cursor.execute("DELETE FROM transactions WHERE id = ?", (tx["id"],))

    cursor.execute("DELETE FROM invoices WHERE id = ?", (invoice_id,))
    conn.commit()
    conn.close()
    return {"success": True, "message": "Đã xóa hóa đơn"}

File: frontend/js/invoices.js
// Invoices, Meter Readings & VietQR Billing Module
const invoicesModule = {
    allInvoices: [],
    currentInvoice: null,
    suggestData: null,

    async init() {
        await this.loadInvoices();
        this.bindEvents();
    },

    async loadInvoices() {
        const month = document.getElementById('filterInvMonth')?.value || '';
        const status = document.getElementById('filterInvStatus')?.value || '';
        const roomId = document.getElementById('filterInvRoom')?.value || '';

        const params = {};
        if (month) params.month = month;
        if (status) params.payment_status = status;
        if (roomId) params.room_id = roomId;

        try {
            const list = await api.getInvoices(params);
            this.allInvoices = list;
            this.renderTable(list);
            this.renderStats(list);
        } catch (error) {
            window.showToast('Lỗi khi tải danh sách hóa đơn: ' + error.message, 'error');
        }
    },

    renderTable(list) {
        const tbody = document.getElementById('invTableBody');
        if (!tbody) return;

        if (list.length === 0) {
            tbody.innerHTML = `
                <tr>
                    <td colspan="7" style="text-align:center; padding:36px; color:var(--text-dim);">
                        <i class="fa-solid fa-file-invoice-dollar" style="font-size:32px; margin-bottom:10px; display:block;"></i>
                        Chưa có hóa đơn nào cho tháng được chọn.
                    </td>
                </tr>
            `;
            return;
        }

        tbody.innerHTML = list.map(inv => {
            const isPaid = inv.payment_status === 'paid';
            const statusBadge = isPaid
                ? '<span class="badge badge-success"><i class="fa-solid fa-check-circle"></i> Đã thanh toán</span>'
                : '<span class="badge badge-warning"><i class="fa-solid fa-clock"></i> Chưa thu tiền</span>';

            return `
                <tr>
                    <td>
                        <strong>${inv.billing_month}</strong>
                        <div style="font-size:11.5px; color:var(--text-dim);">Lập: ${inv.billing_date}</div>
                    </td>
                    <td>
                        <strong style="font-size:15px; color:var(--primary);">${inv.room_code}</strong>
                        <div style="font-size:12px; color:var(--text-muted);">${inv.tenant_name} (${inv.tenant_phone || '-'})</div>
                    </td>
                    <td>
                        <div style="font-size:12.5px;">⚡ Điện: <strong>${inv.electricity_usage} kWh</strong> (${window.formatCurrency(inv.electricity_amount)})</div>
                        <div style="font-size:12.5px;">💧 Nước: <strong>${inv.water_usage || '-'}</strong> (${window.formatCurrency(inv.water_amount)})</div>
                    </td>
                    <td>
                        <div style="font-size:12.5px;">Phòng: ${window.formatCurrency(inv.room_amount)}</div>
                        <div style="font-size:11.5px; color:var(--text-dim);">Dịch vụ: ${window.formatCurrency(inv.internet_amount + inv.garbage_amount + inv.parking_amount + inv.other_amount)}</div>
                    </td>
                    <td>
                        <strong style="font-size:15px; color:var(--text-main);">${window.formatCurrency(inv.total_amount)}</strong>
                    </td>
                    <td>${statusBadge}</td>
                    <td style="text-align:right;">
                        <div style="display:flex; gap:6px; justify-content:flex-end;">
                            <button class="btn btn-outline" style="padding:6px 10px; font-size:12px;" onclick="invoicesModule.openBillPreviewModal(${inv.id})">
                                <i class="fa-solid fa-receipt"></i> Xem Phiếu/QR
                            </button>
                            ${!isPaid ? `
                                <button class="btn btn-success" style="padding:6px 10px; font-size:12px;" onclick="invoicesModule.openPayModal(${inv.id})">
                                    <i class="fa-solid fa-check"></i> Thu tiền
                                </button>
                            ` : ''}
                            <button class="btn-icon" style="width:30px; height:30px;" title="Xóa hóa đơn" onclick="invoicesModule.deleteInvoice(${inv.id})">
                                <i class="fa-solid fa-trash" style="font-size:12px; color:var(--danger);"></i>
                            </button>
                        </div>
                    </td>
                </tr>
            `;
        }).join('');
    },

    renderStats(list) {
        let totalBilled = 0;
        let totalCollected = 0;
        let totalUnpaid = 0;

        list.forEach(inv => {
            totalBilled += inv.total_amount;
            if (inv.payment_status === 'paid') totalCollected += inv.paid_amount || inv.total_amount;
            else totalUnpaid += inv.total_amount;
        });

        const billedEl = document.getElementById('statInvTotalBilled');
        const collectedEl = document.getElementById('statInvCollected');
        const unpaidEl = document.getElementById('statInvUnpaid');

        if (billedEl) billedEl.textContent = window.formatCurrency(totalBilled);
        if (collectedEl) collectedEl.textContent = window.formatCurrency(totalCollected);
        if (unpaidEl) unpaidEl.textContent = window.formatCurrency(totalUnpaid);
    },

    // --- Modal: Chốt điện nước nhanh ---
    async openQuickBillModal(roomId) {
        try {
            const data = await api.getInvoiceSuggest(roomId);
            this.suggestData = data;

            document.getElementById('billRoomName').textContent = `${data.room_code} (Tầng ${data.floor})`;
            document.getElementById('billTenantName').textContent = `${data.tenant_name} (${data.tenant_phone})`;
            document.getElementById('billMonthInput').value = data.billing_month;
            document.getElementById('billDateInput').value = data.billing_date;

            document.getElementById('billPrevElec').value = data.prev_electricity;
            document.getElementById('billCurrElec').value = data.prev_electricity;
            document.getElementById('billElecRate').value = data.electricity_rate;

            document.getElementById('billWaterType').value = data.water_billing_type;
            document.getElementById('billPrevWater').value = data.prev_water;
            document.getElementById('billCurrWater').value = data.prev_water;
            document.getElementById('billWaterRate').value = data.water_rate;
            document.getElementById('billWaterCount').value = data.water_person_count;

            document.getElementById('billRoomAmount').value = data.room_amount;
            document.getElementById('billInternetAmount').value = data.internet_amount;
            document.getElementById('billGarbageAmount').value = data.garbage_amount;
            document.getElementById('billParkingAmount').value = data.parking_amount;
            document.getElementById('billOtherAmount').value = data.other_amount;
            document.getElementById('billOtherNote').value = data.other_note || '';
            document.getElementById('billDiscountAmount').value = 0;

            this.calculateLiveBill();
            document.getElementById('modalQuickBill').classList.add('active');
        } catch (error) {
            window.showToast(error.message, 'error');
        }
    },

    calculateLiveBill() {
        const prevElec = parseFloat(document.getElementById('billPrevElec')?.value) || 0;
        const currElec = parseFloat(document.getElementById('billCurrElec')?.value) || 0;
        const elecRate = parseFloat(document.getElementById('billElecRate')?.value) || 3500;
        const elecUsage = Math.max(0, currElec - prevElec);
        const elecAmount = elecUsage * elecRate;

        const waterType = document.getElementById('billWaterType')?.value || 'per_person';
        const prevWater = parseFloat(document.getElementById('billPrevWater')?.value) || 0;
        const currWater = parseFloat(document.getElementById('billCurrWater')?.value) || 0;
        const waterRate = parseFloat(document.getElementById('billWaterRate')?.value) || 30000;
        const waterCount = parseInt(document.getElementById('billWaterCount')?.value) || 1;

        let waterAmount = 0;
        let waterUsageText = '';
        if (waterType === 'meter') {
            const waterUsage = Math.max(0, currWater - prevWater);
            waterAmount = waterUsage * waterRate;
            waterUsageText = `${waterUsage} m³`;
        } else if (waterType === 'per_person') {
            waterAmount = waterCount * waterRate;
            waterUsageText = `${waterCount} người`;
        } else {
            waterAmount = waterRate;
            waterUsageText = 'Cố định';
        }

        const roomAmount = parseFloat(document.getElementById('billRoomAmount')?.value) || 0;
        const internet = parseFloat(document.getElementById('billInternetAmount')?.value) || 0;
        const garbage = parseFloat(document.getElementById('billGarbageAmount')?.value) || 0;
        const parking = parseFloat(document.getElementById('billParkingAmount')?.value) || 0;
        const other = parseFloat(document.getElementById('billOtherAmount')?.value) || 0;
        const discount = parseFloat(document.getElementById('billDiscountAmount')?.value) || 0;

        const total = roomAmount + elecAmount + waterAmount + internet + garbage + parking + other - discount;

        // Update preview labels
        const elecSummary = document.getElementById('billElecLiveSummary');
        const waterSummary = document.getElementById('billWaterLiveSummary');
        const totalSummary = document.getElementById('billTotalLiveSummary');

        if (elecSummary) elecSummary.textContent = `${elecUsage} kWh = ${window.formatCurrency(elecAmount)}`;
        if (waterSummary) waterSummary.textContent = `${waterUsageText} = ${window.formatCurrency(waterAmount)}`;
        if (totalSummary) totalSummary.textContent = window.formatCurrency(total);

        return {
            elecUsage,
            elecAmount,
            waterAmount,
            total
        };
    },

    async handleSaveInvoice(e) {
        e.preventDefault();
        if (!this.suggestData) return;

        const calc = this.calculateLiveBill();
        const payload = {
            contract_id: this.suggestData.contract_id,
            room_id: this.suggestData.room_id,
            billing_month: document.getElementById('billMonthInput').value,
            billing_date: document.getElementById('billDateInput').value,
            prev_electricity: parseFloat(document.getElementById('billPrevElec').value) || 0,
            curr_electricity: parseFloat(document.getElementById('billCurrElec').value) || 0,
            electricity_usage: calc.elecUsage,
            electricity_rate: parseFloat(document.getElementById('billElecRate').value) || 3500,
            electricity_amount: calc.elecAmount,
            water_billing_type: document.getElementById('billWaterType').value,
            prev_water: parseFloat(document.getElementById('billPrevWater').value) || 0,
            curr_water: parseFloat(document.getElementById('billCurrWater').value) || 0,
            water_usage: 0,
            water_rate: parseFloat(document.getElementById('billWaterRate').value) || 30000,
            water_amount: calc.waterAmount,
            room_amount: parseFloat(document.getElementById('billRoomAmount').value) || 0,
            internet_amount: parseFloat(document.getElementById('billInternetAmount').value) || 0,
            garbage_amount: parseFloat(document.getElementById('billGarbageAmount').value) || 0,
            parking_amount: parseFloat(document.getElementById('billParkingAmount').value) || 0,
            other_amount: parseFloat(document.getElementById('billOtherAmount').value) || 0,
            other_note: document.getElementById('billOtherNote').value.trim(),
            discount_amount: parseFloat(document.getElementById('billDiscountAmount').value) || 0,
            total_amount: calc.total,
            payment_status: 'unpaid',
            paid_amount: 0,
            paid_date: null,
            wallet_id: null,
            notes: 'Lập hóa đơn tiền trọ tháng'
        };

        try {
            const newInv = await api.createInvoice(payload);
            window.showToast('Đã tạo hóa đơn tiền trọ thành công!', 'success');
            document.getElementById('modalQuickBill').classList.remove('active');
            await this.loadInvoices();
            await window.app.refreshGlobalStats();
            
            // Open bill preview immediately
            this.openBillPreviewModal(newInv.id);
        } catch (error) {
            window.showToast(error.message, 'error');
        }
    },

    // --- Modal: Xem Hóa đơn / Bill Zalo / VietQR ---
    async openBillPreviewModal(invoiceId) {
        try {
            const inv = await api.getInvoice(invoiceId);
            this.currentInvoice = inv;

            // Tìm ví ngân hàng mặc định để sinh QR
            const wallets = window.transactionsModule?.allWallets || await api.getWallets();
            const bankWallet = wallets.find(w => w.type === 'bank' && w.account_number) || wallets.find(w => w.type === 'bank') || wallets[0];

            let qrUrl = '';
            let bankInfoText = '';
            if (bankWallet && bankWallet.account_number) {
                const bankCode = bankWallet.bank_code || 'MB';
                const accNum = bankWallet.account_number;
                const accName = bankWallet.account_holder || 'CHU NHA';
                const memo = `Tien tro ${inv.room_code} thang ${inv.billing_month}`;
                qrUrl = `https://img.vietqr.io/image/${bankCode}-${accNum}-compact2.png?amount=${Math.round(inv.total_amount)}&addInfo=${encodeURIComponent(memo)}&accountName=${encodeURIComponent(accName)}`;
                bankInfoText = `${bankWallet.bank_name || bankCode} - STK: ${accNum} (${accName})`;
            }

            const body = document.getElementById('billPreviewBody');
            body.innerHTML = `
                <div class="invoice-bill-card" id="printableBill">
                    <div class="bill-header">
                        <div class="bill-title">PHIẾU THU TIỀN PHÒNG TRỌ</div>
                        <div class="bill-sub">Tháng: <strong>${inv.billing_month}</strong> | Ngày lập: ${inv.billing_date}</div>
                    </div>

                    <div class="bill-info-grid">
                        <div>Phòng: <strong style="color:#4f46e5; font-size:15px;">${inv.room_code} (Tầng ${inv.floor})</strong></div>
                        <div>Khách thuê: <strong>${inv.tenant_name}</strong></div>
                        <div>Số điện thoại: <strong>${inv.tenant_phone || '-'}</strong></div>
                        <div>Trạng thái: <strong style="color:${inv.payment_status === 'paid' ? '#10b981' : '#f59e0b'};">${inv.payment_status === 'paid' ? 'ĐÃ THANH TOÁN' : 'CHƯA THANH TOÁN'}</strong></div>
                    </div>

                    <table class="bill-table">
                        <thead>
                            <tr>
                                <th>Khoản mục</th>
                                <th style="text-align:center;">Chi tiết</th>
                                <th style="text-align:right;">Thành tiền</th>
                            </tr>
                        </thead>
                        <tbody>
                            <tr>
                                <td><strong>1. Tiền phòng</strong></td>
                                <td style="text-align:center;">1 tháng</td>
                                <td style="text-align:right;"><strong>${window.formatCurrency(inv.room_amount)}</strong></td>
                            </tr>
                            <tr>
                                <td><strong>2. Tiền điện</strong></td>
                                <td style="text-align:center;">${inv.prev_electricity} ➜ ${inv.curr_electricity} (${inv.electricity_usage} kWh x ${window.formatCurrency(inv.electricity_rate)})</td>
                                <td style="text-align:right;"><strong>${window.formatCurrency(inv.electricity_amount)}</strong></td>
                            </tr>
                            <tr>
                                <td><strong>3. Tiền nước</strong></td>
                                <td style="text-align:center;">${inv.water_billing_type === 'meter' ? `${inv.water_usage} m³` : 'Khoán'}</td>
                                <td style="text-align:right;"><strong>${window.formatCurrency(inv.water_amount)}</strong></td>
                            </tr>
                            ${inv.internet_amount > 0 ? `
                                <tr>
                                    <td>4. Mạng Wifi</td>
                                    <td style="text-align:center;">Cố định</td>
                                    <td style="text-align:right;">${window.formatCurrency(inv.internet_amount)}</td>
                                </tr>
                            ` : ''}
                            ${inv.garbage_amount > 0 ? `
                                <tr>
                                    <td>5. Rác & Vệ sinh</td>
                                    <td style="text-align:center;">Cố định</td>
                                    <td style="text-align:right;">${window.formatCurrency(inv.garbage_amount)}</td>
                                </tr>
                            ` : ''}
                            ${inv.parking_amount > 0 ? `
                                <tr>
                                    <td>6. Phí gửi xe</td>
                                    <td style="text-align:center;">Cố định</td>
                                    <td style="text-align:right;">${window.formatCurrency(inv.parking_amount)}</td>
                                </tr>
                            ` : ''}
                            ${inv.other_amount > 0 ? `
                                <tr>
                                    <td>7. Dịch vụ khác (${inv.other_note || 'Khác'})</td>
                                    <td style="text-align:center;">-</td>
                                    <td style="text-align:right;">${window.formatCurrency(inv.other_amount)}</td>
                                </tr>
                            ` : ''}
                            ${inv.discount_amount > 0 ? `
                                <tr>
                                    <td style="color:#f43f5e;">8. Giảm trừ</td>
                                    <td style="text-align:center;">Khuyến mãi</td>
                                    <td style="text-align:right; color:#f43f5e;">-${window.formatCurrency(inv.discount_amount)}</td>
                                </tr>
                            ` : ''}
                            <tr class="total-row">
                                <td colspan="2">TỔNG CỘNG THANH TOÁN:</td>
                                <td style="text-align:right;" class="bill-total-highlight">${window.formatCurrency(inv.total_amount)}</td>
                            </tr>
                        </tbody>
                    </table>

                    ${qrUrl ? `
                        <div class="vietqr-section">
                            <div style="font-weight:700; font-size:13px; color:#1e293b;">QUÉT MÃ VIETQR ĐỂ CHUYỂN KHOẢN NHANH</div>
                            <img src="${qrUrl}" alt="VietQR" class="vietqr-img" />
                            <div class="vietqr-tip">${bankInfoText}</div>
                        </div>
                    ` : ''}
                </div>
            `;

            document.getElementById('btnCopyZaloText').onclick = () => this.copyZaloBillText(inv, bankInfoText);
            document.getElementById('modalBillPreview').classList.add('active');
        } catch (error) {
            window.showToast('Không thể tải chi tiết hóa đơn: ' + error.message, 'error');
        }
    },

    copyZaloBillText(inv, bankInfo) {
        const text = `🏡 THÔNG BÁO TIỀN PHÒNG THÁNG ${inv.billing_month}
Phòng: ${inv.room_code} - Khách: ${inv.tenant_name}
---------------------------------
1. Tiền phòng: ${window.formatCurrency(inv.room_amount)}
2. Tiền điện: ${inv.prev_electricity} -> ${inv.curr_electricity} (${inv.electricity_usage} kWh x ${window.formatCurrency(inv.electricity_rate)}) = ${window.formatCurrency(inv.electricity_amount)}
3. Tiền nước: ${window.formatCurrency(inv.water_amount)}
4. Wifi + Rác + Dịch vụ: ${window.formatCurrency(inv.internet_amount + inv.garbage_amount + inv.parking_amount + inv.other_amount)}
${inv.discount_amount > 0 ? `Giảm trừ: -${window.formatCurrency(inv.discount_amount)}\n` : ''}---------------------------------
👉 TỔNG CỘNG: ${window.formatCurrency(inv.total_amount)}

${bankInfo ? `💳 Thông tin thanh toán:\n${bankInfo}\nNội dung: Tien tro ${inv.room_code} thang ${inv.billing_month}` : 'Vui lòng thanh toán cho quản lý tòa nhà.'}
Cảm ơn bạn!`;

        navigator.clipboard.writeText(text).then(() => {
            window.showToast('Đã sao chép nội dung tin nhắn Zalo!', 'success');
        }).catch(() => {
            window.showToast('Không thể tự sao chép, hãy chọn và copy thủ công', 'warning');
        });
    },

    // --- Modal: Thu tiền / Thanh toán hóa đơn ---
    openPayModal(invoiceId) {
        const inv = this.allInvoices.find(i => i.id === invoiceId);
        if (!inv) return;
        this.currentInvoice = inv;

        document.getElementById('payInvoiceRoomName').textContent = `${inv.room_code} (Tháng ${inv.billing_month})`;
        document.getElementById('payInvoiceAmount').textContent = window.formatCurrency(inv.total_amount);
        document.getElementById('payDateInput').value = new Date().toISOString().split('T')[0];

        // Fill wallets select
        const wallets = (window.walletsModule?.allWallets && window.walletsModule.allWallets.length > 0)
            ? window.walletsModule.allWallets 
            : (window.transactionsModule?.allWallets || []);
        const walletSelect = document.getElementById('payWalletSelect');
        walletSelect.innerHTML = wallets.map(w => `<option value="${w.id}">${w.name} (${window.formatCurrency(w.balance)})</option>`).join('');

        document.getElementById('modalPayInvoice').classList.add('active');
    },

    async handleConfirmPayment(e) {
        e.preventDefault();
        if (!this.currentInvoice) return;

        const wallet_id = parseInt(document.getElementById('payWalletSelect').value);
        const paid_date = document.getElementById('payDateInput').value;

        try {
            await api.payInvoice(this.currentInvoice.id, {
                payment_status: 'paid',
                paid_amount: this.currentInvoice.total_amount,
                paid_date,
                wallet_id,
                auto_create_income: true
            });

            window.showToast('Đã xác nhận thanh toán & đồng bộ vào Sổ Thu Chi!', 'success');
            document.getElementById('modalPayInvoice').classList.remove('active');
            await this.loadInvoices();
            await window.transactionsModule?.loadTransactions();
            await window.app.refreshGlobalStats();
        } catch (error) {
            window.showToast(error.message, 'error');
        }
    },

    async deleteInvoice(id) {
        if (!confirm('Bạn có chắc muốn xóa hóa đơn này? Nếu đã thanh toán, giao dịch thu nhập sẽ được hủy hoàn tiền.')) return;
        try {
            await api.deleteInvoice(id);
            window.showToast('Đã xóa hóa đơn', 'success');
            await this.loadInvoices();
            await window.transactionsModule?.loadTransactions();
            await window.app.refreshGlobalStats();
        } catch (error) {
            window.showToast(error.message, 'error');
        }
    },

    bindEvents() {
        const formBill = document.getElementById('formQuickBill');
        if (formBill) formBill.addEventListener('submit', (e) => this.handleSaveInvoice(e));

        const formPay = document.getElementById('formPayInvoice');
        if (formPay) formPay.addEventListener('submit', (e) => this.handleConfirmPayment(e));

        // Live calculation listeners
        ['billCurrElec', 'billPrevElec', 'billElecRate', 'billCurrWater', 'billPrevWater', 'billWaterRate', 'billWaterCount', 'billRoomAmount', 'billInternetAmount', 'billGarbageAmount', 'billParkingAmount', 'billOtherAmount', 'billDiscountAmount', 'billWaterType'].forEach(id => {
            const el = document.getElementById(id);
            if (el) {
                el.addEventListener('input', () => this.calculateLiveBill());
                el.addEventListener('change', () => this.calculateLiveBill());
            }
        });

        ['filterInvMonth', 'filterInvStatus', 'filterInvRoom'].forEach(id => {
            const el = document.getElementById(id);
            if (el) el.addEventListener('change', () => this.loadInvoices());
        });
    }
};

window.invoicesModule = invoicesModule;
